File: backend/app/schemas/market.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime
from decimal import Decimal

class OHLCVResponse(BaseModel):
    id: str
    symbol: str
    exchange: str
    interval: str
    timestamp: datetime
    open: Decimal
    high: Decimal
    low: Decimal
    close: Decimal
    volume: Decimal
    vwap: Optional[Decimal]
    number_of_trades: Optional[int]

    class Config:
        from_attributes = True

class CurrentPriceResponse(BaseModel):
    symbol: str
    price: Decimal
    bid: Optional[Decimal]
    ask: Optional[Decimal]
    volume_24h: Optional[Decimal]
    high_24h: Optional[Decimal]
    low_24h: Optional[Decimal]
    updated_at: datetime

    class Config:
        from_attributes = True

class PriceHistoryRequest(BaseModel):
    symbol: str
    interval: str = "1h"  # 1m, 5m, 15m, 1h, 4h, 1d, 1w
    limit: int = 100
    start_time: Optional[datetime] = None
    end_time: Optional[datetime] = None

class SymbolInfo(BaseModel):
    symbol: str
    base_asset: str
    quote_asset: str
    min_quantity: Decimal
    max_quantity: Decimal
    tick_size: Decimal
    min_notional: Optional[Decimal]

File: frontend/src/api/market.ts
import { apiClient } from './client';

export interface MarketPrice {
  symbol: string;
  price: number;
  change24h: number;
  volume24h: number;
  high24h: number;
  low24h: number;
  timestamp: string;
}

export interface OHLCV {
  timestamp: number;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
}

export interface SymbolInfo {
  symbol: string;
  baseAsset: string;
  quoteAsset: string;
  minSize: number;
  maxSize: number;
  stepSize: number;
  tickSize: number;
}

export interface PriceUpdate {
  symbol: string;
  price: number;
  change24h: number;
  volume: number;
}

export const marketApi = {
  // 📊 Prices
  getPrices: () =>
    apiClient.get<MarketPrice[]>('/market/prices'),
  
  getPrice: (symbol: string) =>
    apiClient.get<MarketPrice>(`/market/price/${symbol}`),
  
  // 📈 OHLCV Data
  getOHLCV: (symbol: string, interval: '1m' | '5m' | '15m' | '1h' | '4h' | '1d' | '1w', limit?: number) =>
    apiClient.get<OHLCV[]>(`/market/ohlcv/${symbol}`, { 
      params: { interval, limit } 
    }),
  
  refreshOHLCV: (symbol: string, interval: string, limit?: number) =>
    apiClient.post(`/market/ohlcv/refresh`, null, { 
      params: { symbol, interval, limit } 
    }),
  
  // 📋 Symbols
  getSymbols: () =>
    apiClient.get<string[]>('/market/symbols'),
  
  getSymbolInfo: (symbol: string) =>
    apiClient.get<SymbolInfo>(`/market/symbols/${symbol}`),
  
  // 🔄 Real-time Updates
  getLatestUpdates: () =>
    apiClient.get<PriceUpdate[]>('/market/updates'),
  
  refreshPrices: () =>
    apiClient.post('/market/prices/refresh'),
  
  // 📊 Market Status
  getMarketStatus: () =>
    apiClient.get<{
      isOpen: boolean;
      nextClose?: string;
      nextOpen?: string;
    }>('/market/status'),
};

export type { OHLCV, MarketPrice, SymbolInfo, PriceUpdate };